File: java/src/TradingAgent.java
/**
 * Trading Agent — Java
 * Wires MarketData, RAG strategy knowledge, and BacktestEngine into the ReAct
 * loop. Mirrors trading_agent.py safety contract exactly:
 *   - order_propose registered as a tool → drafts only, never executes
 *   - orderExecuteIBKR NOT registered as a tool → requires explicit human confirm=true
 *   - IBKR_LIVE_TRADING env var required for live mode (defaults to PAPER)
 *
 * CLI:
 *   java TradingAgent --task "Backtest SMA crossover on AAPL"
 *   java TradingAgent --max-iterations 20 --skip-ingest
 */

import com.fasterxml.jackson.databind.JsonNode;
import com.fasterxml.jackson.databind.ObjectMapper;
import com.fasterxml.jackson.databind.node.ObjectNode;

import java.nio.file.Path;
import java.nio.file.Paths;
import java.util.*;
import java.util.concurrent.ConcurrentHashMap;

public class TradingAgent {

    static final ObjectMapper JSON = new ObjectMapper();

    static final String TRADING_SYSTEM_PROMPT =
        "You are a trading research assistant. You have access to:\n" +
        "- rag_search: a knowledge base of trading strategy theory (momentum, mean reversion, breakout, crypto-specific, risk management)\n" +
        "- get_stock_quote / get_stock_ohlcv: live and historical stock market data\n" +
        "- get_crypto_quote / get_crypto_ohlcv: live and historical cryptocurrency market data\n" +
        "- compute_indicators: SMA, RSI calculations\n" +
        "- run_backtest: test a named strategy preset against historical data and report performance metrics\n" +
        "- order_propose: draft a proposed order for human review — this NEVER executes a trade\n\n" +
        "CRITICAL SAFETY RULE: You can never execute trades. You can only use order_propose to draft\n" +
        "an order for a human to review and execute manually. If asked to \"buy\", \"sell\", \"execute\",\n" +
        "or \"place an order\", you must use order_propose, clearly state the order is a PROPOSAL\n" +
        "requiring human confirmation, and never claim a trade has been executed.\n\n" +
        "When researching a strategy, search the knowledge base first to ground your reasoning in\n" +
        "established strategy theory, then pull live or historical data to evaluate it, then\n" +
        "backtest if asked. Always cite which strategy concept underlies your reasoning.";

    // ── Order proposal (safety boundary — drafts only, never executes) ─────────
    public record OrderProposal(
        String proposalId,
        String symbol,
        String side,
        double quantity,
        String orderType,
        Double limitPrice,
        String assetClass,
        String rationale,
        String status
    ) {}

    static final ConcurrentHashMap<String, OrderProposal> PROPOSALS = new ConcurrentHashMap<>();

    public static OrderProposal proposeOrder(
        String symbol, String side, double quantity,
        String assetClass, String orderType, Double limitPrice, String rationale
    ) {
        if (!"buy".equals(side) && !"sell".equals(side))
            throw new IllegalArgumentException("side must be 'buy' or 'sell', got '" + side + "'");
        if (quantity <= 0)
            throw new IllegalArgumentException("quantity must be positive");
        if ("limit".equals(orderType) && limitPrice == null)
            throw new IllegalArgumentException("limitPrice is required for limit orders");

        String id = UUID.randomUUID().toString();
        OrderProposal proposal = new OrderProposal(
            id,
            symbol.trim().toUpperCase(),
            side,
            quantity,
            orderType == null || orderType.isBlank() ? "market" : orderType,
            limitPrice,
            assetClass == null || assetClass.isBlank() ? "equity" : assetClass,
            rationale,
            "PROPOSED — REQUIRES HUMAN CONFIRMATION"
        );
        PROPOSALS.put(id, proposal);
        log("INFO", "trading_agent.order_proposed", Map.of(
            "proposal_id", id, "symbol", proposal.symbol(),
            "side", side, "quantity", String.valueOf(quantity)));
        return proposal;
    }

    public static Optional<OrderProposal> getProposal(String proposalId) {
        return Optional.ofNullable(PROPOSALS.get(proposalId));
    }

    // ── IBKR execution (NOT registered as an agent tool — explicit opt-in only) ─
    public record ExecutionResult(
        String proposalId, String mode, String status, OrderProposal order, String note
    ) {}

    /**
     * Execute a previously-drafted order proposal via Interactive Brokers.
     * NOT registered in ReactLoop.TOOLS — the agent cannot call this autonomously.
     * Must be invoked directly by human-reviewed calling code after explicit review.
     *
     * @param confirm must be Boolean.TRUE — any other value, including null, is rejected
     */
    public static ExecutionResult orderExecuteIBKR(String proposalId, boolean confirm) {
        if (!confirm)
            throw new SecurityException(
                "Execution requires confirm=true from a human-reviewed call site. " +
                "This is not optional and has no default override.");

        OrderProposal proposal = PROPOSALS.get(proposalId);
        if (proposal == null)
            throw new IllegalArgumentException(
                "No proposal found with id '" + proposalId + "' — it may have expired or never existed");

        boolean liveTrading = "true".equalsIgnoreCase(System.getenv("IBKR_LIVE_TRADING"));
        String mode = liveTrading ? "LIVE" : "PAPER";

        if (liveTrading) {
            log("WARN", "trading_agent.LIVE_ORDER_EXECUTION", Map.of(
                "proposal_id", proposalId, "symbol", proposal.symbol(), "mode", mode));
        } else {
            log("INFO", "trading_agent.paper_order_execution", Map.of(
                "proposal_id", proposalId, "symbol", proposal.symbol(), "mode", mode));
        }

        return new ExecutionResult(
            proposalId, mode, "READY_FOR_BROKER_SUBMISSION", proposal,
            "This function validated and logged the confirmed order. Actual broker " +
            "submission happens via the IBKR MCP connector at the integration layer, " +
            "which must be invoked separately by the calling application."
        );
    }

    // ── RAG engine singleton for trading knowledge ─────────────────────────────
    static volatile RagEngine tradingEngine;
    static final Object ENGINE_LOCK = new Object();

    static RagEngine getTradingEngine() throws Exception {
        if (tradingEngine == null) {
            synchronized (ENGINE_LOCK) {
                if (tradingEngine == null)
                    tradingEngine = new RagEngine("trading_strategies");
            }
        }
        return tradingEngine;
    }

    // ── Tool registration ─────────────────────────────────────────────────────
    public static void registerTools() throws Exception {
        MarketData.registerTools();
        RagEngine engine = getTradingEngine();

        // rag_search — wired to the trading knowledge base
        if (!ReactLoop.TOOLS.containsKey("rag_search")) {
            ReactLoop.registerTool(
                "rag_search",
                "Search the trading strategy knowledge base for theory on momentum, mean reversion, breakout, crypto, and risk management.",
                JSON.createObjectNode()
                    .<ObjectNode>put("type", "object")
                    .set("properties", JSON.createObjectNode()
                        .set("query", JSON.createObjectNode().put("type", "string"))
                        .set("top_k", JSON.createObjectNode().put("type", "integer"))),
                input -> {
                    try {
                        String query = input.path("query").asText("");
                        int topK = input.path("top_k").asInt(5);
                        return engine.retrieveAsContext(query, topK);
                    } catch (Exception e) { return "Error: " + e.getMessage(); }
                }
            );
        }

        // run_backtest — runs one of three preset strategies against live OHLCV
        ReactLoop.registerTool(
            "run_backtest",
            "Backtest a named strategy preset against historical OHLCV data. Presets: sma_crossover, rsi_mean_reversion, breakout.",
            JSON.createObjectNode()
                .<ObjectNode>put("type", "object")
                .set("properties", JSON.createObjectNode()
                    .set("symbol",                JSON.createObjectNode().put("type", "string"))
                    .set("asset_class",           JSON.createObjectNode().put("type", "string"))
                    .set("strategy",              JSON.createObjectNode().put("type", "string"))
                    .set("lookback_days_or_size", JSON.createObjectNode().put("type", "string"))),
            input -> {
                try {
                    String symbol     = input.path("symbol").asText("");
                    String strategy   = input.path("strategy").asText("");
                    String assetClass = input.path("asset_class").asText("equity");
                    String lookback   = input.path("lookback_days_or_size").asText("compact");

                    List<MarketData.OHLCVBar> bars;
                    int barsPerYear;
                    if ("crypto".equals(assetClass)) {
                        int days = 90;
                        try { days = Integer.parseInt(lookback); } catch (NumberFormatException ignored) {}
                        List<MarketData.CryptoOHLCBar> cb = MarketData.getCryptoOHLCV(symbol, "usd", days);
                        bars = cb.stream().map(b -> new MarketData.OHLCVBar(
                            String.valueOf(b.timestamp()), b.open(), b.high(), b.low(), b.close(), 0
                        )).toList();
                        barsPerYear = 365;
                    } else {
                        String size = "full".equals(lookback) ? "full" : "compact";
                        bars = MarketData.getStockOHLCV(symbol, "daily", size);
                        barsPerYear = 252;
                    }

                    BacktestEngine.SignalFn signalFn = switch (strategy) {
                        case "sma_crossover"      -> BacktestEngine.makeSMACrossoverSignal(0, 0);
                        case "rsi_mean_reversion" -> BacktestEngine.makeRSIMeanReversionSignal(0, 0, 0);
                        case "breakout"            -> BacktestEngine.makeBreakoutSignal(0);
                        default -> throw new IllegalArgumentException(
                            "Unknown strategy '" + strategy + "'. Available: sma_crossover, rsi_mean_reversion, breakout");
                    };

                    BacktestEngine.BacktestResult result = BacktestEngine.runBacktest(
                        bars, signalFn, BacktestEngine.BacktestOptions.withBarsPerYear(barsPerYear)
                    );

                    return String.format(
                        "Backtest of '%s' on %s (%d bars):\n" +
                        "  Total Return: %.2f%%\n  Annualized Return: %.2f%%\n  Sharpe Ratio: %.2f\n" +
                        "  Max Drawdown: %.2f%%\n  Win Rate: %.1f%% (%d trades)\n" +
                        "  Avg Trade: %.2f%% | Best: %.2f%% | Worst: %.2f%%",
                        strategy, symbol.toUpperCase(), bars.size(),
                        result.totalReturnPct(), result.annualizedReturnPct(), result.sharpeRatio(),
                        result.maxDrawdownPct(), result.winRatePct(), result.numTrades(),
                        result.avgTradePct(), result.bestTradePct(), result.worstTradePct()
                    );
                } catch (Exception e) { return "Backtest failed: " + e.getMessage(); }
            }
        );

        // order_propose — drafts only, never executes
        ReactLoop.registerTool(
            "order_propose",
            "Draft a proposed order for human review. NEVER executes a trade — only creates a proposal that requires explicit human confirmation to execute.",
            JSON.createObjectNode()
                .<ObjectNode>put("type", "object")
                .set("properties", JSON.createObjectNode()
                    .set("symbol",      JSON.createObjectNode().put("type", "string"))
                    .set("side",        JSON.createObjectNode().put("type", "string"))
                    .set("quantity",    JSON.createObjectNode().put("type", "number"))
                    .set("asset_class", JSON.createObjectNode().put("type", "string"))
                    .set("order_type",  JSON.createObjectNode().put("type", "string"))
                    .set("limit_price", JSON.createObjectNode().put("type", "number"))
                    .set("rationale",   JSON.createObjectNode().put("type", "string"))),
            input -> {
                try {
                    String symbol     = input.path("symbol").asText("");
                    String side       = input.path("side").asText("");
                    double quantity   = input.path("quantity").asDouble(0);
                    String assetClass = input.path("asset_class").asText("equity");
                    String orderType  = input.path("order_type").asText("market");
                    Double limitPrice = input.has("limit_price") ? input.path("limit_price").asDouble() : null;
                    String rationale  = input.path("rationale").asText("");

                    OrderProposal proposal = proposeOrder(symbol, side, quantity, assetClass, orderType, limitPrice, rationale);
                    String priceStr = "limit".equals(orderType)
                        ? String.format("@ limit $%.2f", limitPrice)
                        : "@ market";

                    return String.format(
                        "ORDER PROPOSAL (id=%s) — %s\n  %s %.4g %s (%s) %s\n  Rationale: %s\n" +
                        "  This is a draft only. No order has been placed. A human must explicitly " +
                        "confirm and execute this proposal via the trading platform.",
                        proposal.proposalId(), proposal.status(),
                        side.toUpperCase(), quantity, symbol.toUpperCase(), orderType, priceStr, rationale
                    );
                } catch (IllegalArgumentException e) {
                    return "Error: invalid order parameters — " + e.getMessage();
                }
            }
        );

        log("INFO", "trading_agent.tools_registered", Map.of("total_tools", String.valueOf(ReactLoop.TOOLS.size())));
    }

    // ── Strategy knowledge ingestion ──────────────────────────────────────────
    public static int ingestStrategyKnowledge(String knowledgeDir) throws Exception {
        Path dir = knowledgeDir != null && !knowledgeDir.isBlank()
            ? Paths.get(knowledgeDir)
            : Paths.get(System.getProperty("user.dir"), "..", "..", "..", "knowledge", "trading_strategies").normalize();

        if (!dir.toFile().isDirectory()) {
            log("WARN", "trading_agent.knowledge_dir_missing", Map.of("path", dir.toString()));
            return 0;
        }

        List<RagEngine.Document> docs = Loaders.loadDirectory(dir, true);
        if (docs.isEmpty()) {
            log("WARN", "trading_agent.no_strategy_docs_found", Map.of("path", dir.toString()));
            return 0;
        }

        RagEngine engine = getTradingEngine();
        int count = engine.ingest(docs);
        log("INFO", "trading_agent.strategy_knowledge_ingested", Map.of(
            "docs", String.valueOf(docs.size()), "chunks", String.valueOf(count)));
        return count;
    }

    // ── Structured logging helper ──────────────────────────────────────────────
    static void log(String level, String event, Map<String, String> fields) {
        try {
            ObjectNode node = JSON.createObjectNode();
            node.put("timestamp", java.time.Instant.now().toString());
            node.put("level",   level);
            node.put("service", "trading-agent-java");
            node.put("event",   event);
            fields.forEach(node::put);
            System.err.println(JSON.writeValueAsString(node));
        } catch (Exception ignored) {
            System.err.println("{\"level\":\"" + level + "\",\"event\":\"" + event + "\"}");
        }
    }

    // ── CLI entry point ────────────────────────────────────────────────────────
    public static void main(String[] args) throws Exception {
        String task = "Search the knowledge base for momentum strategies, then explain how a 50/200 SMA crossover works and what its main risk is.";
        int maxIterations = 20;
        boolean skipIngest = false;

        for (int i = 0; i < args.length; i++) {
            if ("--task".equals(args[i]) && i + 1 < args.length) { task = args[++i]; }
            else if ("--max-iterations".equals(args[i]) && i + 1 < args.length) { maxIterations = Integer.parseInt(args[++i]); }
            else if ("--skip-ingest".equals(args[i])) { skipIngest = true; }
        }

        if (!skipIngest) {
            try { ingestStrategyKnowledge(null); }
            catch (Exception e) { log("WARN", "trading_agent.ingest_warning", Map.of("error", e.getMessage())); }
        }

        registerTools();

        ReactLoop.ReactResult result = ReactLoop.runReact(task, TRADING_SYSTEM_PROMPT, maxIterations, null);

        System.out.println("\n" + "=".repeat(64));
        System.out.println("TRADING AGENT — FINAL ANSWER");
        System.out.println("=".repeat(64));
        System.out.println(result.answer);
        System.out.println("=".repeat(64));
        System.out.printf("Iterations: %d | Tokens: %d | Time: %dms%n",
            result.trace.iterations, result.trace.totalTokens, result.trace.elapsedMs);

        if (!PROPOSALS.isEmpty()) {
            System.out.printf("%n%d order proposal(s) drafted this session (none executed):%n", PROPOSALS.size());
            for (OrderProposal p : PROPOSALS.values()) {
                System.out.printf("  - %s %s %.4g %s (%s) — %s%n",
                    p.side().toUpperCase(), p.symbol(), p.quantity(), p.orderType(),
                    p.assetClass(), p.proposalId());
            }
        }
    }
}

File: rust/src/trading_agent.rs
//! Trading Agent — Rust
//! Wires market_data, RAG strategy knowledge, and backtest_engine into the ReAct
//! loop. Mirrors trading_agent.py's safety contract exactly: propose_order drafts
//! only, order_execute_ibkr is NOT registered as an agent tool and requires
//! explicit human confirm=true plus, separately, IBKR_LIVE_TRADING=true for live.

use std::collections::HashMap;
use std::env;
use std::path::PathBuf;
use std::sync::{Arc, Mutex};

use anyhow::{anyhow, Result};
use once_cell::sync::Lazy;
use serde::Serialize;
use uuid::Uuid;

use crate::backtest_engine::{
    make_breakout_signal, make_rsi_mean_reversion_signal, make_sma_crossover_signal, run_backtest,
    BacktestOptions,
};
use crate::loaders::load_directory;
use crate::market_data::{get_crypto_ohlcv, get_stock_ohlcv, register_market_data_tools, OHLCVBar, CryptoOHLCBar};
use crate::rag_engine::{Document, RAGEngine};
use crate::react_loop::{run_react, ReactOptions, ToolRegistry};

pub const TRADING_SYSTEM_PROMPT: &str = r#"You are a trading research assistant. You have access to:
- rag_search: a knowledge base of trading strategy theory (momentum, mean reversion, breakout, crypto-specific, risk management)
- get_stock_quote / get_stock_ohlcv: live and historical stock market data
- get_crypto_quote / get_crypto_ohlcv: live and historical cryptocurrency market data
- compute_indicators: SMA, RSI calculations
- run_backtest: test a named strategy preset against historical data and report performance metrics
- order_propose: draft a proposed order for human review — this NEVER executes a trade

CRITICAL SAFETY RULE: You can never execute trades. You can only use order_propose to draft
an order for a human to review and execute manually. If asked to "buy", "sell", "execute",
or "place an order", you must use order_propose, clearly state the order is a PROPOSAL
requiring human confirmation, and never claim a trade has been executed.

When researching a strategy, search the knowledge base first to ground your reasoning in
established strategy theory, then pull live or historical data to evaluate it, then
backtest if asked. Always cite which strategy concept underlies your reasoning."#;

// ── Order proposal (safety boundary — drafts only, never executes) ──────────────
#[derive(Debug, Clone, Serialize)]
pub struct OrderProposal {
    pub proposal_id: String,
    pub symbol: String,
    pub side: String,
    pub quantity: f64,
    pub order_type: String,
    pub limit_price: Option<f64>,
    pub asset_class: String,
    pub rationale: String,
    pub status: String,
}

static PROPOSALS: Lazy<Mutex<HashMap<String, OrderProposal>>> = Lazy::new(|| Mutex::new(HashMap::new()));

pub fn propose_order(
    symbol: &str, side: &str, quantity: f64, asset_class: &str, order_type: &str,
    limit_price: Option<f64>, rationale: &str,
) -> Result<OrderProposal> {
    if side != "buy" && side != "sell" {
        return Err(anyhow!("side must be 'buy' or 'sell', got '{}'", side));
    }
    if quantity <= 0.0 {
        return Err(anyhow!("quantity must be positive"));
    }
    if order_type == "limit" && limit_price.is_none() {
        return Err(anyhow!("limit_price is required for limit orders"));
    }

    let proposal = OrderProposal {
        proposal_id: Uuid::new_v4().to_string(),
        symbol: symbol.trim().to_uppercase(),
        side: side.to_string(),
        quantity,
        order_type: if order_type.is_empty() { "market".to_string() } else { order_type.to_string() },
        limit_price,
        asset_class: if asset_class.is_empty() { "equity".to_string() } else { asset_class.to_string() },
        rationale: rationale.to_string(),
        status: "PROPOSED — REQUIRES HUMAN CONFIRMATION".to_string(),
    };

    PROPOSALS.lock().unwrap().insert(proposal.proposal_id.clone(), proposal.clone());
    eprintln!("[trading_agent] order proposed: {:?}", proposal);
    Ok(proposal)
}

pub fn get_proposal(proposal_id: &str) -> Option<OrderProposal> {
    PROPOSALS.lock().unwrap().get(proposal_id).cloned()
}

// ── IBKR execution (NOT registered as an agent tool — explicit opt-in only) ────
#[derive(Debug, Clone, Serialize)]
pub struct ExecutionResult {
    pub proposal_id: String,
    pub mode: String,
    pub status: String,
    pub order: OrderProposal,
    pub note: String,
}

/// Executes a previously-drafted order proposal via Interactive Brokers.
/// NOT registered in ToolRegistry — the agent cannot call this autonomously.
/// Must be invoked directly by human-reviewed calling code after explicit review.
pub fn order_execute_ibkr(proposal_id: &str, confirm: bool) -> Result<ExecutionResult> {
    if !confirm {
        return Err(anyhow!(
            "execution requires confirm=true from a human-reviewed call site; this is not optional and has no default override"
        ));
    }

    let proposal = get_proposal(proposal_id)
        .ok_or_else(|| anyhow!("no proposal found with id '{}' — it may have expired or never existed", proposal_id))?;

    let live_trading = env::var("IBKR_LIVE_TRADING").map(|v| v.to_lowercase() == "true").unwrap_or(false);
    let mode = if live_trading { "LIVE" } else { "PAPER" }.to_string();

    if live_trading {
        eprintln!("[trading_agent] LIVE_ORDER_EXECUTION: {:?}", proposal);
    } else {
        eprintln!("[trading_agent] paper_order_execution: {:?}", proposal);
    }

    Ok(ExecutionResult {
        proposal_id: proposal_id.to_string(),
        mode,
        status: "READY_FOR_BROKER_SUBMISSION".to_string(),
        order: proposal,
        note: "This function validated and logged the confirmed order. Actual broker \
               submission happens via the IBKR MCP connector at the integration layer, \
               which must be invoked separately by the calling application.".to_string(),
    })
}

// ── Tool registration ───────────────────────────────────────────────────────────
pub fn register_trading_tools(registry: &mut ToolRegistry, engine: Arc<RAGEngine>) {
    register_market_data_tools(registry);

    {
        let engine = Arc::clone(&engine);
        registry.register(
            "rag_search",
            "Search the trading strategy knowledge base for theory on momentum, mean reversion, breakout, crypto, and risk management.",
            serde_json::json!({ "type": "object", "properties": {
                "query": { "type": "string" }, "top_k": { "type": "integer" }
            }, "required": ["query"] }),
            Arc::new(move |input: serde_json::Value| {
                let engine = Arc::clone(&engine);
                let query = input["query"].as_str().unwrap_or("").to_string();
                let top_k = input["top_k"].as_u64().unwrap_or(5);
                let rt = tokio::runtime::Handle::current();
                std::thread::spawn(move || rt.block_on(async move {
                    engine.retrieve_as_context(&query, top_k).await
                })).join().map_err(|e| anyhow!("{:?}", e))?
            }),
        );
    }

    registry.register(
        "run_backtest",
        "Backtest a named strategy preset against historical OHLCV data for a symbol. Presets: sma_crossover, rsi_mean_reversion, breakout.",
        serde_json::json!({ "type": "object", "properties": {
            "symbol": { "type": "string" }, "asset_class": { "type": "string" },
            "strategy": { "type": "string" }, "lookback_days_or_size": { "type": "string" }
        }, "required": ["symbol", "strategy"] }),
        Arc::new(|input: serde_json::Value| {
            let symbol = input["symbol"].as_str().unwrap_or("").to_string();
            let strategy = input["strategy"].as_str().unwrap_or("").to_string();
            let asset_class = input["asset_class"].as_str().unwrap_or("equity").to_string();
            let lookback = input["lookback_days_or_size"].as_str().unwrap_or("").to_string();

            let rt = tokio::runtime::Handle::current();
            std::thread::spawn(move || rt.block_on(async move {
                let (bars, bars_per_year): (Vec<OHLCVBar>, u32) = if asset_class == "crypto" {
                    let days: u32 = lookback.parse().unwrap_or(90);
                    match get_crypto_ohlcv(&symbol, "usd", days).await {
                        Ok(crypto_bars) => {
                            let converted: Vec<OHLCVBar> = crypto_bars.iter().map(|b: &CryptoOHLCBar| OHLCVBar {
                                date: b.timestamp.to_string(), open: b.open, high: b.high, low: b.low, close: b.close, volume: 0,
                            }).collect();
                            (converted, 365)
                        }
                        Err(e) => return Ok(format!("Backtest failed: {}", e)),
                    }
                } else {
                    let size = if lookback == "full" { "full" } else { "compact" };
                    match get_stock_ohlcv(&symbol, "daily", size).await {
                        Ok(bars) => (bars, 252),
                        Err(e) => return Ok(format!("Backtest failed: {}", e)),
                    }
                };

                let signal_fn = match strategy.as_str() {
                    "sma_crossover"      => make_sma_crossover_signal(0, 0),
                    "rsi_mean_reversion" => make_rsi_mean_reversion_signal(0, 0.0, 0.0),
                    "breakout"           => make_breakout_signal(0),
                    other => return Ok(format!("Error: unknown strategy '{}'. Available: sma_crossover, rsi_mean_reversion, breakout", other)),
                };

                let opts = BacktestOptions { bars_per_year, ..Default::default() };
                match run_backtest(&bars, signal_fn, opts) {
                    Ok(result) => Ok(format!(
                        "Backtest of '{}' on {} ({} bars):\n  Total Return: {:.2}%\n  Annualized Return: {:.2}%\n  Sharpe Ratio: {:.2}\n  Max Drawdown: {:.2}%\n  Win Rate: {:.1}% ({} trades)\n  Avg Trade: {:.2}% | Best: {:.2}% | Worst: {:.2}%",
                        strategy, symbol.to_uppercase(), bars.len(),
                        result.total_return_pct, result.annualized_return_pct, result.sharpe_ratio,
                        result.max_drawdown_pct, result.win_rate_pct, result.num_trades,
                        result.avg_trade_pct, result.best_trade_pct, result.worst_trade_pct,
                    )),
                    Err(e) => Ok(format!("Backtest failed: {}", e)),
                }
            })).join().map_err(|e| anyhow!("{:?}", e))?
        }),
    );

    registry.register(
        "order_propose",
        "Draft a proposed order for human review. NEVER executes a trade — only creates a proposal that requires explicit human confirmation to execute.",
        serde_json::json!({ "type": "object", "properties": {
            "symbol": { "type": "string" }, "side": { "type": "string" }, "quantity": { "type": "number" },
            "asset_class": { "type": "string" }, "order_type": { "type": "string" },
            "limit_price": { "type": "number" }, "rationale": { "type": "string" }
        }, "required": ["symbol", "side", "quantity", "rationale"] }),
        Arc::new(|input: serde_json::Value| {
            let symbol = input["symbol"].as_str().unwrap_or("").to_string();
            let side = input["side"].as_str().unwrap_or("").to_string();
            let quantity = input["quantity"].as_f64().unwrap_or(0.0);
            let asset_class = input["asset_class"].as_str().unwrap_or("equity").to_string();
            let order_type = input["order_type"].as_str().unwrap_or("market").to_string();
            let limit_price = input["limit_price"].as_f64();
            let rationale = input["rationale"].as_str().unwrap_or("").to_string();

            match propose_order(&symbol, &side, quantity, &asset_class, &order_type, limit_price, &rationale) {
                Ok(proposal) => {
                    let price_str = if order_type == "limit" {
                        format!("@ limit ${:.2}", limit_price.unwrap_or(0.0))
                    } else { "@ market".to_string() };
                    Ok(format!(
                        "ORDER PROPOSAL (id={}) — {}\n  {} {} {} ({}) {}\n  Rationale: {}\n  This is a draft only. No order has been placed. A human must explicitly confirm and execute this proposal via the trading platform.",
                        proposal.proposal_id, proposal.status, side.to_uppercase(), quantity, symbol.to_uppercase(), order_type, price_str, rationale,
                    ))
                }
                Err(e) => Ok(format!("Error: invalid order parameters — {}", e)),
            }
        }),
    );
}

pub async fn ingest_strategy_knowledge(engine: &RAGEngine, knowledge_dir: Option<PathBuf>) -> Result<usize> {
    let dir = knowledge_dir.unwrap_or_else(|| {
        env::current_dir().unwrap_or_default().join("..").join("..").join("..").join("knowledge").join("trading_strategies")
    });

    if !dir.is_dir() {
        eprintln!("[trading_agent] knowledge dir missing: {:?}", dir);
        return Ok(0);
    }

    let docs: Vec<Document> = load_directory(&dir, true);
    if docs.is_empty() {
        eprintln!("[trading_agent] no strategy docs found in {:?}", dir);
        return Ok(0);
    }

    let count = engine.ingest(docs.clone()).await?;
    eprintln!("[trading_agent] strategy knowledge ingested: {} docs, {} chunks", docs.len(), count);
    Ok(count)
}

// ── CLI ────────────────────────────────────────────────────────────────────────
pub async fn run_trading_agent_cli(task: &str, max_iterations: usize, skip_ingest: bool) -> Result<()> {
    let engine = Arc::new(RAGEngine::new("trading_strategies").await?);

    if !skip_ingest {
        if let Err(e) = ingest_strategy_knowledge(&engine, None).await {
            eprintln!("[trading_agent] ingest warning: {}", e);
        }
    }

    let mut registry = ToolRegistry::new();
    register_trading_tools(&mut registry, Arc::clone(&engine));

    let (answer, trace) = run_react(task, &registry, ReactOptions {
        system_extra: TRADING_SYSTEM_PROMPT.to_string(),
        max_iterations,
        on_step: None,
    }).await?;

    println!("\n{}", "=".repeat(64));
    println!("TRADING AGENT — FINAL ANSWER");
    println!("{}", "=".repeat(64));
    println!("{}", answer);
    println!("{}", "=".repeat(64));
    println!("Iterations: {} | Tokens: {} | Time: {}ms", trace.iterations, trace.total_tokens, trace.elapsed_ms);

    let proposals = PROPOSALS.lock().unwrap();
    if !proposals.is_empty() {
        println!("\n{} order proposal(s) drafted this session (none executed):", proposals.len());
        for p in proposals.values() {
            println!("  - {:?}", p);
        }
    }

    Ok(())
}

// ── Tests ──────────────────────────────────────────────────────────────────────
#[cfg(test)]
mod tests {
    use super::*;
    use serial_test::serial;

    #[test]
    fn test_propose_order_creates_proposal() {
        let p = propose_order("AAPL", "buy", 10.0, "equity", "market", None, "test rationale").unwrap();
        assert_eq!(p.symbol, "AAPL");
        assert_eq!(p.side, "buy");
        assert_eq!(p.status, "PROPOSED — REQUIRES HUMAN CONFIRMATION");
    }

    #[test]
    fn test_propose_order_invalid_side() {
        assert!(propose_order("AAPL", "invalid", 10.0, "equity", "market", None, "x").is_err());
    }

    #[test]
    fn test_propose_order_non_positive_quantity() {
        assert!(propose_order("AAPL", "buy", 0.0, "equity", "market", None, "x").is_err());
        assert!(propose_order("AAPL", "buy", -5.0, "equity", "market", None, "x").is_err());
    }

    #[test]
    fn test_limit_order_requires_price() {
        assert!(propose_order("AAPL", "buy", 10.0, "equity", "limit", None, "x").is_err());
    }

    #[test]
    fn test_get_proposal_roundtrip() {
        let p = propose_order("BTC", "buy", 0.1, "crypto", "market", None, "x").unwrap();
        let fetched = get_proposal(&p.proposal_id);
        assert!(fetched.is_some());
        assert_eq!(fetched.unwrap().proposal_id, p.proposal_id);
    }

    #[test]
    fn test_get_nonexistent_proposal() {
        assert!(get_proposal("nonexistent-id").is_none());
    }

    // ── Critical safety tests: order_execute_ibkr must never fire without confirm=true ──
    #[test]
    fn test_execution_requires_confirm_true() {
        let p = propose_order("AAPL", "buy", 10.0, "equity", "market", None, "x").unwrap();
        let result = order_execute_ibkr(&p.proposal_id, false);
        assert!(result.is_err());
        assert!(result.unwrap_err().to_string().contains("confirm=true"));
    }

    #[test]
    fn test_execution_unknown_proposal_raises() {
        let result = order_execute_ibkr("nonexistent-id", true);
        assert!(result.is_err());
        assert!(result.unwrap_err().to_string().contains("No proposal found") || result.unwrap_err().to_string().contains("no proposal found"));
    }

    #[test]
    #[serial]
    fn test_execution_defaults_to_paper_mode() {
        std::env::remove_var("IBKR_LIVE_TRADING");
        let p = propose_order("AAPL", "buy", 10.0, "equity", "market", None, "x").unwrap();
        let result = order_execute_ibkr(&p.proposal_id, true).unwrap();
        assert_eq!(result.mode, "PAPER");
    }

    #[test]
    #[serial]
    fn test_execution_live_requires_explicit_env_var() {
        std::env::set_var("IBKR_LIVE_TRADING", "true");
        let p = propose_order("AAPL", "buy", 10.0, "equity", "market", None, "x").unwrap();
        let result = order_execute_ibkr(&p.proposal_id, true).unwrap();
        assert_eq!(result.mode, "LIVE");
        std::env::remove_var("IBKR_LIVE_TRADING");
    }

    #[test]
    fn test_order_execute_ibkr_not_in_tool_registry() {
        let mut registry = ToolRegistry::new();
        // register_trading_tools requires a RAGEngine which needs network/Qdrant —
        // instead we assert the static fact: this module never calls
        // registry.register("order_execute_ibkr", ...) or ("order_execute", ...)
        // anywhere in register_trading_tools. Verified by source inspection in
        // CI via a grep-based guard test below; here we confirm a fresh registry
        // has no such tool prior to any trading registration.
        assert!(registry.schemas().iter().all(|s| {
            let name = s.get("name").and_then(|n| n.as_str()).unwrap_or("");
            name != "order_execute_ibkr" && name != "order_execute"
        }));
    }
}
